### PR TITLE
update source code for Python 3.5, sdpython/pythonnet

### PR DESCRIPTION
**Pull request by vmuriart**
PR to merge in sdpython's work. 

+ Whitespace changes have been removed to make it easier to see changes
+ Cleaned the history from un-related changes

From:
http://www.xavierdupre.fr/app/pymyinstall/helpsphinx/blog/2016/2016-01-03_pythonnet.html
https://github.com/sdpython/pythonnet

Needs review and testing

### CODE REVIEW

**Comment by vmuriart** [comment]
@tonyroberts cleaned out the code to make it easier to review. There are [some changes](https://github.com/vmuriart/pythonnet/commit/e6becfe066b289e0a5981b7c1450652b6c42e8f5#diff-f34eee2812824d797e6464eca8cd3e15L335) that I think might break something. Also, looks like Travis/AppVeyor don't want to tests due to outdated pip ?

@sdpython is there a particular reason those were removed?

**Comment by denfromufa** [comment]
@vmuriart have you run test case before and after you changes? there are very few failing, like this one:

https://github.com/pythonnet/pythonnet/issues/122

**Comment by vmuriart** [comment]
Haven't, just created a pr for the python3.5 library without changing his code. I'm about to start testing and dissecting this a bit more but wanted to get some feedback on some of the changes from @tonyroberts and @sdpython

**Comment by vmuriart** [comment]
I think i misunderstood you. Sounds like I rebased ontop of a failing code.

**Comment by denfromufa** [comment]
@vmuriart can you point to problems you are having?

**Comment by tonyroberts** [comment]
@vmuriart I don't see any obvious problems with this PR, thanks for getting this done. I'll sort out the CI builds over the weekend or early next week and then we can see how it looks.

**Comment by denfromufa** [comment]
@tonyroberts are unit tests running as part of CI builds?

**Comment by tonyroberts** [comment]
@denfromufa yes, that's pretty much the whole point of running the CI builds.

**Comment by denfromufa** [comment]
@tonyroberts, very weird I see that the tests do not fail on build machine, locally on my machine, but if I run using PTVS GUI, then some tests fail.

Look like Python.EmbeddingTest not running in CI?

**Comment by denfromufa** [comment]
OK I see the problem: runtests.py skips 10 tests.

**Comment by tonyroberts** [comment]
I've fixed the CI builds in the develop branch, and pulled this PR into my fork to test. Looks like there are some problems with the Python 2 builds and not all tests are passing for 3.5 yet.

https://ci.appveyor.com/project/TonyRoberts/pythonnet/build/2.0.129
https://travis-ci.org/tonyroberts/pythonnet/builds/107813141

File: src/runtime/assemblymanager.cs [hunk 1]
```diff
@@ -185,7 +185,7 @@ internal class AssemblyManager {
             try {
                 assembly = Assembly.Load(name);
             }
-            catch (System.Exception e) {
+            catch (System.Exception /*e*/) {
```
Comment (denfromufa) [inline src/runtime/assemblymanager.cs]: @tonyroberts according to Jon Skeet http://stackoverflow.com/a/8009801/2230844 it is possible to "miss" some unmanaged exceptions when `catch` is used without arguments.

However I have not tested this hypothesis myself, it just looked the only weird place to me.

**Comment by denfromufa** [comment]
@tonyroberts @vmuriart this branch is using old appveyor yaml config.

**Comment by denfromufa** [comment]
@tonyroberts @vmuriart I fixed one bug with @sdpython code for python 3.5 support which passes 2.7 tests (failed before), but still getting segfault on 3.5 tests. I'm also working on getting the correct the appveyor yaml config, since python 3.5 switched from .msi to .exe installer. all this happens here:

https://github.com/denfromufa/pythonnet/tree/develop

**Comment by denfromufa** [comment]
Here is appveyor guideline how to setup python testing with tox:

http://python-packaging-user-guide.readthedocs.org/en/latest/appveyor/

Also some examples from @ogrisel

https://github.com/ogrisel/python-appveyor-demo

**Comment by denfromufa** [comment]
@tonyroberts I have advanced in my branch and only failing some tests on python 3.5 travis-ci. Appveyor is still a mess due to switch from msi to exe.

https://github.com/denfromufa/pythonnet/tree/develop

**Comment by denfromufa** [comment]
@vmuriart @tonyroberts 

feel free to grab this branch https://github.com/denfromufa/pythonnet/tree/develop

I only commented out one weird test line:

https://github.com/denfromufa/pythonnet/blob/develop/src/tests/test_array.py#L146

every other test is passing!

File: src/runtime/assemblymanager.cs (hunk 1 above)
Comment (denfromufa) [inline src/runtime/assemblymanager.cs]: my hypothesis was wrong

**Comment by vmuriart** [comment]
sorry been out for a while w work. 
@denfromufa Did you want to submit a pr, and i can close mine? Do you mean its passing on all versions?

**Comment by denfromufa** [comment]
@vmuriart yes, all tests on travis ci are passing.

https://travis-ci.org/denfromufa/pythonnet

 I carried over your branch, so you can close this pull request.
I'm trying to figure out how to add python 3.5 on appveyor since the installer changed from .msi to .exe. but this yml/cmd/ps salad is not familiar to me so I asked for appveyor support:

http://help.appveyor.com/discussions/problems/4167-error-parsing-appveyoryml-value-can-be-either-string-or-secure-string-line-26-column-6

**Pull request closed without merging**